File: batch_extract.js
const fs = require('fs');

// Parse fields from SAP SE11 browser snapshot
function parseFieldsFromSnapshot(snapshot, beschreibung) {
  const fields = [];

  // Match all row patterns in the grid
  // Format: row "Zum Auswählen einer Zeile drücken Sie die Leertaste. FELDNAME   DATENELEMENT DATENTYP LAENGE DEZST 0 KURZTEXT":
  const lines = snapshot.split('\n');

  for (const line of lines) {
    // Match the row pattern
    const rowMatch = line.match(/row "Zum Auswählen.*?\. (\S+)\s+(\S+)\s+(\S+)\s+(\d+)\s+(\d+)\s+\d+\s+([^"]+)"/);
    if (rowMatch) {
      const [, feldname, datenelement, datentyp, laenge, dezst, kurztext] = rowMatch;

      // Check if key field (has [checked] in the row context)
      // We need to look for the checkbox pattern after the field name
      const isKey = line.includes('[checked]');

      fields.push({
        feldname,
        datenelement,
        datentyp,
        laenge: parseInt(laenge, 10),
        dezimalstellen: parseInt(dezst, 10),
        kurztext: kurztext.trim(),
        schluesselfeld: isKey
      });
    }
  }

  return {
    beschreibung,
    anzahl_felder: fields.length,
    felder: fields
  };
}

// Read existing metadata
function loadMetadata() {
  try {
    const content = fs.readFileSync('us4g_metadata.json', 'utf8');
    return JSON.parse(content);
  } catch (e) {
    return {
      extraktionsdatum: new Date().toISOString().split('T')[0],
      quelle: "SAP SE11 - ABAP Dictionary",
      beschreibung: "S/4 Utilities /US4G/ Tabellen-Feldmetadaten",
      fortschritt: { extrahiert: 0, gesamt: 613 },
      tabellen: {}
    };
  }
}

// Save metadata
function saveMetadata(meta) {
  fs.writeFileSync('us4g_metadata.json', JSON.stringify(meta, null, 2), 'utf8');
}

// Add table to metadata
function addTable(tableName, tableData) {
  const meta = loadMetadata();
  meta.tabellen[tableName] = tableData;
  meta.fortschritt.extrahiert = Object.keys(meta.tabellen).length;
  saveMetadata(meta);
  return meta.fortschritt.extrahiert;
}

module.exports = { parseFieldsFromSnapshot, loadMetadata, saveMetadata, addTable };

// Test
if (require.main === module) {
  const testSnapshot = `
  - textbox "Kurzbeschreibung": Test Description
  - row "Zum Auswählen einer Zeile drücken Sie die Leertaste. MANDT   MANDT CLNT 3 0 0 Mandant":
    - gridcell "Zum Auswählen einer Zeile drücken Sie die Leertaste."
    - gridcell "MANDT"
    - gridcell "": [checked]
  `;
  const result = parseFieldsFromSnapshot(testSnapshot, "Test");
  console.log(JSON.stringify(result, null, 2));
}

File: batch_extract.py
#!/usr/bin/env python3
"""Helper script to parse SE11 snapshot and update metadata JSON."""
import re
import json
import sys


def parse_se11_snapshot(snapshot: str) -> dict | None:
    """Parse SE11 browser_snapshot and return table metadata."""
    # Check if table doesn't exist
    if "existiert nicht" in snapshot or "does not exist" in snapshot.lower():
        return None

    result = {
        "beschreibung": "",
        "anzahl_felder": 0,
        "felder": []
    }

    # Extract description
    desc_match = re.search(r'textbox "Kurzbeschreibung": ([^\n]+)', snapshot)
    if desc_match:
        result["beschreibung"] = desc_match.group(1).strip()

    # Extract field count from pagination
    count_match = re.search(r'button /\s*\n\s*- button "(\d+)"', snapshot)
    if count_match:
        result["anzahl_felder"] = int(count_match.group(1))

    # Parse field rows
    field_pattern = r'row ["\']Zum Auswählen[^"\']*Leertaste\. (\S+)\s+(\S+)\s+(\w+)\s+(\d+)\s+(\d+)\s+\d+\s+([^"\']+)["\']:'

    # Split by rows to check key fields
    rows = re.split(r'(?=row ["\']Zum Auswählen)', snapshot)

    for row_text in rows:
        match = re.search(field_pattern, row_text)
        if match:
            is_key = '[checked]' in row_text.split('checkbox', 1)[1] if 'checkbox' in row_text else False

            field = {
                "feldname": match.group(1),
                "datenelement": match.group(2),
                "datentyp": match.group(3),
                "laenge": int(match.group(4)),
                "dezimalstellen": int(match.group(5)),
                "kurztext": match.group(6).strip(),
                "schluesselfeld": is_key
            }
            result["felder"].append(field)

    if not result["anzahl_felder"]:
        result["anzahl_felder"] = len(result["felder"])

    return result


def update_metadata_file(metadata_path: str, table_name: str, table_data: dict):
    """Add table data to metadata JSON file."""
    with open(metadata_path, 'r', encoding='utf-8') as f:
        metadata = json.load(f)

    metadata["tabellen"][table_name] = table_data
    metadata["fortschritt"]["extrahiert"] = len(metadata["tabellen"])

    with open(metadata_path, 'w', encoding='utf-8') as f:
        json.dump(metadata, f, ensure_ascii=False, indent=2)


if __name__ == "__main__":
    if len(sys.argv) < 3:
        print("Usage: python batch_extract.py METADATA_PATH TABLE_NAME < snapshot.txt")
        sys.exit(1)

    metadata_path = sys.argv[1]
    table_name = sys.argv[2]
    snapshot = sys.stdin.read()

    result = parse_se11_snapshot(snapshot)
    if result is None:
        print(json.dumps({"status": "not_found", "table": table_name}))
    else:
        update_metadata_file(metadata_path, table_name, result)
        print(json.dumps({
            "status": "ok",
            "table": table_name,
            "fields_found": len(result["felder"]),
            "total_fields": result["anzahl_felder"]
        }))
